### PR TITLE
coprocessor/dag/expr: add some un-pushed builtin UDFs(Interval)

### PR DESCRIPTION
**Pull request by bb7133**
## What have you changed? (mandatory)

implemented ScalarFuncSig::IntervalInt and ScalarFuncSig::IntervalReal

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

added some UTs

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

https://github.com/pingcap/tikv/issues/3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_compare.rs [hunk 1]
```diff
@@ -202,6 +202,68 @@ impl ScalarFunc {
...
+
+        let mut left = 1;
+        let mut right = self.children.len();
+        while left < right {
+            let mid = left + (right - left) / 2;
+            let m = match self.children[mid].eval_int(ctx, row) {
+                Err(e) => return Err(e),
+                Ok(None) => target,
+                Ok(Some(v)) => v,
+            };
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_compare.rs]: Could we use `self.children[mid].eval_int(ctx,row)?.unwrap_or(target);` instead?

**Comment by AndreMouche** [comment]
@breeswish PTAL

File: src/coprocessor/dag/expr/builtin_compare.rs (hunk 1 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/builtin_compare.rs]: That would be awesome, thank you!

File: src/coprocessor/dag/expr/builtin_compare.rs [hunk 2]
```diff
@@ -202,6 +202,60 @@ impl ScalarFunc {
...
+                left = mid + 1
+            } else {
+                right = mid
+            }
+        }
+        Ok(Some((left - 1) as i64))
+    }
+
+    pub fn interval_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = match self.children[0].eval_real(ctx, row) {
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_compare.rs]: might be written as:

```rust
let target = match self.children[0].eval_real(ctx, row)? {
    None => ...
    Some(v) => ...
};
```

File: src/coprocessor/dag/expr/builtin_compare.rs [hunk 3]
```diff
@@ -202,6 +202,60 @@ impl ScalarFunc {
...
+            Err(e) => return Err(e),
+            Ok(None) => return Ok(Some(-1)),
+            Ok(Some(v)) => v,
+        };
+
+        let mut left = 1;
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_compare.rs]: Why our initial left is 1 while in TiDB i starts from 0?

File: src/coprocessor/dag/expr/builtin_compare.rs [hunk 4]
```diff
@@ -202,6 +202,60 @@ impl ScalarFunc {
     pub fn in_json(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         do_in(self, |v| v.eval_json(ctx, row), |l, r| Ok(l.cmp(r)))
     }
+
+    pub fn interval_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = match self.children[0].eval_int(ctx, row) {
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_compare.rs]: ditto.

File: src/coprocessor/dag/expr/builtin_compare.rs (hunk 2 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/builtin_compare.rs]: Yes, thank you!

File: src/coprocessor/dag/expr/builtin_compare.rs (hunk 3 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/builtin_compare.rs]: The code at TiDB side makes a slice that starts from 1 when passing args to `binSearch`:

https://github.com/pingcap/tidb/blob/cef80b3da390b8b35c703546c54cda9a251a18de/expression/builtin_compare.go#L841

The code here initialize `left` as 1 and return `left - 1` as result, which should be the same as TiDB code. @breeswish

**Comment by bb7133** [comment]
Comments addressed, thank you  @breeswish

**Comment by breeswish** [comment]
@AndreMouche Please give an approval :)

**Pull request merged**